src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+
+    An index is represented by a vector that maps the number of each page to the
+    key stored last in this page or, if the page is an overflow page, to the key
+    of the corresponding key–value pair. From this, the following constraints on
+    the vector follow:
+
+      * The vector must be non-empty.
+
+      * The elements of the vector must be non-decreasing.
+-}
+newtype IndexOrdinary = IndexOrdinary (Vector SerialisedKey)
+
+-- * Search
+
+{-|
+    Returns the span of pages that contain the key–value pair of a certain key
+    if there is such a pair, or an arbitrary, valid page span if there is not.
+-}
+search :: SerialisedKey -> IndexOrdinary -> PageSpan
+search key (IndexOrdinary lastKeys) = PageSpan (PageNo start) (PageNo end) where
+
+    start :: Int
+    start | protoStart < pageCount = protoStart
+          | otherwise              = maybe 0 succ $
+                                     findIndexR (/= lastKeys ! pred pageCount) $
+                                     lastKeys
+
+    end :: Int
+    end | protoStart < pageCount
+            = start + length (takeWhile (== key) (drop (succ start) lastKeys))
+        | otherwise
+            = pred pageCount
+
+    protoStart :: Int
+    protoStart = searchForProtoStart 0 (length lastKeys)
+
+    searchForProtoStart :: Int -> Int -> Int
+    searchForProtoStart first count
+        | count == 0               = first
+        | key <= lastKeys ! center = searchForProtoStart first frontCount
+        | otherwise                = searchForProtoStart (succ center) rearCount
+        where
+
+        frontCount, rearCount, center :: Int
+        frontCount = count `div` 2
+        rearCount  = pred (count - frontCount)
+        center     = first + frontCount
+
+    pageCount :: Int
+    pageCount = length lastKeys
+
+-- * Deserialisation
+
+{-|
+    Reads an index along with the number of entries of the respective run from a
+    byte string.
+
+    The byte string must contain the serialised index exactly, with no leading
+    or trailing space.
+-}
+fromSBS :: ShortByteString -> Either String (NumEntries, IndexOrdinary)
+fromSBS shortByteString@(ShortByteString byteArray)
+    | fullSize < 12
+        = Left "Doesn't contain header and footer"
+    | version == byteSwap32 supportedVersion
+        = Left "Non-matching endianness"
+    | version /= supportedVersion
+        = Left "Unsupported version"
+    | otherwise
+        = (,) <$> entriesCount <*> index
+    where
+
+    fullSize :: Int
+    fullSize = ShortByteString.length shortByteString
+
+    fullBytes :: Primitive.Vector Word8
+    fullBytes = Primitive.Vector 0 fullSize byteArray
+
+    version :: Word32
+    version = indexByteArray byteArray 0
+
+    postVersionBytes :: Primitive.Vector Word8
+    postVersionBytes = Primitive.drop 4 fullBytes
+
+    lastKeysBytes, entriesCountBytes :: Primitive.Vector Word8
+    (lastKeysBytes, entriesCountBytes)
+        = Primitive.splitAt (fullSize - 12) postVersionBytes
```

Comment:
```suggestion
    lastKeysBytes, entriesCountBytes :: Primitive.Vector Word8
    (lastKeysBytes, entriesCountBytes)
        = Primitive.splitAt (fullSize - 12) fullBytes
```

Reply:
I think that my version is the correct one. We only want the payload (`lastKeysBytes`) and the footer (`entriesCountBytes`) and not the header; so we should start with `postVersionBytes`, which we got from `fullBytes` by removing the header. Of course, it seems strange that in the size calculation we then use `fullSize`. However, this is just about numbers, and we do get the size of the payload indeed by subtracting 12 from the full size, because 12 is the total size of the header and footer. Of course, it would be a bit more intuitive to use some `postVersionSize` here, but introducing this seemed overkill to me.